### PR TITLE
[improve][pcip] PCIP-1 Add the development process for pulsar-java-contrib project

### PR DESCRIPTION
**Pull request by StevenLuMT**
Add the development process for pulsar-java-contrib project

### CODE REVIEW

File: pcip/TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
+<!--
+RULES
+* Never place a link to an external site like Google Doc. The proposal should be in this issue entirely.
+* Use a spelling and grammar checker tools if available for you (there are plenty of free ones).
+
+PROPOSAL HEALTH CHECK
+I can read the design document and understand the problem statement and what you plan to change *without* resorting to a couple of hours of code reading just to start having a high level understanding of the change.
+
+IMAGES
+If you need diagrams, avoid attaching large files. You can use [MermaidJS]([url](https://mermaid.js.org/)) as a simple language to describe many types of diagrams.
```
Comment (Denovo1998) [inline pcip/TEMPLATE.md:10]: Images are still needed.

```
IMAGES
If you need the diagrams, please create a folder named pcip-XXX under the pcip/static/img path and put the images in.
```

Comment (StevenLuMT) [reply pcip/TEMPLATE.md]: updated

File: pcip/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+* Bug-fixes
+* Simple enhancements that won't affect the APIs or the semantic
+* Small documentation changes
+* Small website changes
+* Build scripts changes (except: a complete rewrite)
+
+### Who can create a PCIP?
+
+Any person willing to contribute to the Apache Pulsar project is welcome to create a PCIP.
```
Comment (AuroraTwinkle) [inline pcip/README.md:47]: This should be Apache Pulsar Java Contrib project.

File: pcip/TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+* Give a high level explanation on all concepts you will be using throughout this document. For example, if you want to talk about Persistent Subscriptions, explain briefly (1 paragraph) what this is. If you're going to talk about Transaction Buffer, explain briefly what this is. 
+  If you're going to change something specific, then go into more detail about it and how it works. 
+* Provide links where possible if a person wants to dig deeper into the background information. 
+
+DON'T
+* Do not include links *instead* explanation. Do provide links for further explanation.
+
+EXAMPLES
+* See [PCIP-248](https://github.com/apache/pulsar/issues/19601), Background section to get an understanding on how you add the background knowledge needed.
```
Comment (AuroraTwinkle) [inline pcip/TEMPLATE.md:31]: This is a example from pulsar? Maybe we need one from Pulsar Java Contrib

File: pcip/README.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+## How does the PCIP process work?
```
Comment (AuroraTwinkle) [inline pcip/README.md:49]: Since we do'nt need DISCUSSION and VOTE as **Development Process** mentioned, is the PCIP workflow here still necessary?

Comment (StevenLuMT) [reply pcip/README.md:49]: Deleted,Thanks

File: pcip/README.md (hunk 2 above)
Comment (StevenLuMT) [reply pcip/README.md:47]: Updated,Thanks

File: pcip/README.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
+# Pulsar Java Contrib Improvement Proposal (PCIP)
+
+## What is a PCIP?
+
+The PCIP is a "Pulsar Java Contrib Improvement Proposal" and it's the mechanism used to propose changes to the Apache Pulsar Java Contrib codebases.
+
+The changes might be in terms of new features, large code refactoring, changes to APIs.
```
Comment (liangyepianzhou) [inline pcip/README.md:7]: This repository is for collecting new features and new ideas. Will this involve API changes?

File: pcip/README.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+In practical terms, the PCIP defines a process in which developers can submit a design doc, receive feedback and get the "go ahead" to execute.
+
+
+### What is the goal of a PCIP?
+
+There are several goals for the PCIP process:
```
Comment (liangyepianzhou) [inline pcip/README.md:14]: IMO, the goals can be divided into:
1. As a user manual, add instructions when introducing new features or modifying existing features.
2. Explain the functional design ideas to facilitate review and later maintenance.

File: pcip/README.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+Any person willing to contribute to the Apache Pulsar project is welcome to create a PCIP.
+
+## How does the PCIP process work?
+
+A PCIP proposal can be in these states:
+1. **DRAFT**: (Optional) This might be used for contributors to collaborate and to seek feedback on an incomplete version of the proposal.
+
+2. **DISCUSSION**: The proposal has been submitted to the community for feedback and approval.
+
+3. **ACCEPTED**: The proposal has been accepted by the Pulsar project.
```
Comment (liangyepianzhou) [inline pcip/README.md:56]: This should also be the Apache Pulsar Java Contrib project. The same goes for the following.

File: pcip/TEMPLATE.md (hunk 3 above)
Comment (liangyepianzhou) [reply pcip/TEMPLATE.md:31]: +1

File: pcip/README.md (hunk 6 above)
Comment (StevenLuMT) [reply pcip/README.md:14]: done

File: pcip/TEMPLATE.md (hunk 3 above)
Comment (StevenLuMT) [reply pcip/TEMPLATE.md]: done

File: pcip/README.md (hunk 5 above)
Comment (StevenLuMT) [reply pcip/README.md]: Updated

File: pcip/README.md (hunk 7 above)
Comment (StevenLuMT) [reply pcip/README.md]: Deleted

File: pcip/README.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+1. As a user manual, add instructions when introducing new features or modifying existing features. 
+2. Explain the functional design ideas to facilitate review and later maintenance.
+
+It is not a goal for PCIP to add undue process or slow-down the development.
+
+### When is a PCIP required?
+
+* Any new feature for Pulsar Java Contrib
+* Any change to the semantic of existing functionality, even when current behavior is incorrect.
```
Comment (liangyepianzhou) [inline pcip/README.md:24]:
```suggestion
* Any change to the semantic of existing functionality
```

File: pcip/README.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+* Any large code change that will touch multiple components
+* Any change to the configuration
+
+### When is a PCIP *not* required?
+
+* Bug-fixes
+* Small documentation changes
```
Comment (liangyepianzhou) [inline pcip/README.md:31]:
```suggestion
* Documentation changes
```

File: pcip/README.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+* Small website changes
```
Comment (liangyepianzhou) [inline pcip/README.md:32]:
```suggestion
```

File: pcip/README.md [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+* Build scripts changes (except: a complete rewrite)
```
Comment (liangyepianzhou) [inline pcip/README.md:33]:
```suggestion
```

**Comment by AuroraTwinkle** [comment]
LGTM

File: pcip/README.md [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+1. you can put the code changes and design documents(pcip/pcip-xxx.md) in one PR. 
+2. If the code is not yet developed, you can include only a design document(pcip/pcip-xxx.md) in the PR and submit the code separately after the development is completed.
+
+## List of PCIPs
+
+### Historical PCIPs
+You can the view list of PCIPs previously managed by GitHub [here](https://github.com/apache/pulsar-java-contrib/tree/master/pcip)
+
+### List of PCIPs
+1. You can view all PCIPs (besides the historical ones) as the list of Pull Requests having title starting with `[improve][pcip] PCIP-`. Here is the [link](https://github.com/apache/pulsar-java-contrib/pulls?q=is%3Apr+title%3A%22%5Bpcip%5D%5Bdesign%5D+PCIP-%22) for it.
```
Comment (liangyepianzhou) [inline pcip/README.md:61]: I'm thinking we need a PCIP list, which should be maintained in the contributor repository. Then it is best to submit a separate PR to update the PCIP list while the contributor provides PCIP. Add a PCIP whose status is open. The updated PR of this PCIP should be immediately merged by the maintainer without the need for approval by multiple Committers.
Similarly, after PCIP is accepted or rejected, a separate PR should be submitted to update the PCIP list.

Comment (StevenLuMT) [reply pcip/README.md]: You can view all PCIPs (besides the historical ones) as the list of Pull Requests having title starting with `PCIP-`. Here is the [link](https://github.com/apache/pulsar-java-contrib/pulls?q=is%3Apr+%22PCIP-%22+in%3Atitle+sort%3Acreated-desc) for it.

File: pcip/TEMPLATE.md [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+Describe what needs to be considered in Pulsar Geo-Replication in the upgrade and possible downgrade/rollback of this feature.
+-->
+
+# Alternatives
+
+<!--
+If there are alternatives that were already considered by the authors or, after the discussion, by the community, and were rejected, please list them here along with the reason why they were rejected.
+-->
+
+# General Notes
```
Comment (liangyepianzhou) [inline pcip/TEMPLATE.md:157]: Delete it.

File: pcip/TEMPLATE.md [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+
+# Links
+
```
Comment (liangyepianzhou) [inline pcip/TEMPLATE.md:160]: Delete it.

File: pcip/TEMPLATE.md (hunk 13 above)
Comment (StevenLuMT) [reply pcip/TEMPLATE.md]: Done

File: pcip/TEMPLATE.md (hunk 14 above)
Comment (StevenLuMT) [reply pcip/TEMPLATE.md]: done

File: pcip/TEMPLATE.md [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+Specify the list of instructions, if there are such, needed to perform before/after upgrading to Pulsar version containing this feature.
+-->
+
+## Downgrade / Rollback
+
+<!--
+Describe a cookbook detailing the steps required to rollback Pulsar to previous version *without* this feature.
+-->
+
+## Pulsar Geo-Replication Upgrade & Downgrade/Rollback Considerations
```
Comment (liangyepianzhou) [inline pcip/TEMPLATE.md:145]: Delete it

File: pcip/TEMPLATE.md [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+* Response codes, and for each what they mean.
+  For each response code, please include a detailed description of the response body JSON, specifying each field and what it means.
+  This is the place to document the errors.
+-->
+
+### Binary protocol
+
+### Configuration
+
+### CLI
```
Comment (liangyepianzhou) [inline pcip/TEMPLATE.md:110]: IMO, maybe it would be better to replace the following parts with quick startup

File: pcip/TEMPLATE.md (hunk 15 above)
Comment (StevenLuMT) [reply pcip/TEMPLATE.md]: done

File: pcip/TEMPLATE.md (hunk 16 above)
Comment (StevenLuMT) [reply pcip/TEMPLATE.md:108]: Done

**Review by liangyepianzhou — APPROVED** [review]
LGTM

**Comment by Denovo1998** [comment]
LGTM

File: pcip/README.md [hunk 17]
```diff
@@ -0,0 +1,63 @@
+# Pulsar Java Contrib Improvement Proposal (PCIP)
```
Comment (eolivelli) [inline pcip/README.md:1]: This is very close to the PIP process.
Do we have a discussion on the dev ML ?

Comment (StevenLuMT) [reply pcip/README.md:1]: mail disscuss link: https://lists.apache.org/thread/z4bx7207cmprvp0vj8mgf1jhs7q2s4vn

**Review by jiazhai — APPROVED** [review]

**Comment by Jason918** [comment]
LGTM

**Review by Jason918 — APPROVED** [review]
LGTM

**Comment by StevenLuMT** [comment]
Mailing List discussion thread: https://lists.apache.org/thread/z4bx7207cmprvp0vj8mgf1jhs7q2s4vn
Mailing List voting thread: https://lists.apache.org/thread/mmd65whp3g9yd6p5n1ytbqws2gt4cmrl

**Comment by StevenLuMT** [comment]
@liangyepianzhou mailing List vote is passed, help me merge this pr

**Pull request merged**